File: ios-native/DMVCalifornia/Features/Progress/ProgressViewModel.swift
import Foundation

@MainActor
final class ProgressViewModel: ObservableObject {
    @Published var streak = StudyStreak()
    @Published var stats = OverallStats()
    @Published var scoreHistory: [ScoreHistoryEntry] = []
    @Published var weakAreas: [WeakArea] = []
    @Published var bookmarks: [BookmarkedQuestion] = []
    @Published var recentActivity: [QuizAttempt] = []
    @Published var isLoading = false

    private let storageService: StorageServiceProtocol

    init(storageService: StorageServiceProtocol = StorageService.shared) {
        self.storageService = storageService
    }

    func loadData() async {
        isLoading = true

        // Load all data concurrently
        async let streakTask = storageService.getStreak()
        async let statsTask = storageService.getOverallStats()
        async let historyTask = storageService.getScoreHistory(limit: 10)
        async let weakAreasTask = storageService.getWeakAreas()
        async let bookmarksTask = storageService.getBookmarks()
        async let activityTask = storageService.getRecentActivity(limit: 5)

        (streak, stats, scoreHistory, weakAreas, bookmarks, recentActivity) = await (
            streakTask, statsTask, historyTask, weakAreasTask, bookmarksTask, activityTask
        )

        isLoading = false
    }
}

File: android-native/app/src/main/java/com/dmvcalifornia/app/feature/progress/ProgressViewModel.kt
package com.dmvcalifornia.app.feature.progress

import androidx.lifecycle.ViewModel
import com.dmvcalifornia.app.core.data.StorageManager
import com.dmvcalifornia.app.core.model.*
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import javax.inject.Inject

@HiltViewModel
class ProgressViewModel @Inject constructor(
    private val storageManager: StorageManager
) : ViewModel() {

    private val _streak = MutableStateFlow(StudyStreak())
    val streak: StateFlow<StudyStreak> = _streak.asStateFlow()

    private val _stats = MutableStateFlow(OverallStats())
    val stats: StateFlow<OverallStats> = _stats.asStateFlow()

    private val _weakAreas = MutableStateFlow<List<WeakArea>>(emptyList())
    val weakAreas: StateFlow<List<WeakArea>> = _weakAreas.asStateFlow()

    private val _recentActivity = MutableStateFlow<List<QuizAttempt>>(emptyList())
    val recentActivity: StateFlow<List<QuizAttempt>> = _recentActivity.asStateFlow()

    init {
        loadData()
    }

    private fun loadData() {
        _streak.value = storageManager.getStreak()
        _stats.value = storageManager.getOverallStats()
        _weakAreas.value = storageManager.getWeakAreas()
        _recentActivity.value = storageManager.getRecentActivity()
    }
}
